clippy_lints/src/methods/cloned_last.rs

```diff
@@ -0,0 +1,36 @@
+use clippy_utils::diagnostics::{span_lint_and_sugg};
+use clippy_utils::source::snippet;
+use clippy_utils::ty::implements_trait;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::CLONED_LAST;
+
+/// lint use of `cloned().next()` for `Iterators`
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'_>,
+    recv: &'tcx hir::Expr<'_>,
+) {
+    // lint if caller of `.filter().next()` is an Iterator
+    let recv_impls_iterator = cx.tcx.get_diagnostic_item(sym::Iterator).map_or(false, |id| {
+        implements_trait(cx, cx.typeck_results().expr_ty(recv), id, &[])
+    });
+    if recv_impls_iterator {
+        let msg = "called `cloned().last()` on an `Iterator`. It may be more efficient to call
+                   `.last.cloned()` instead";
+        let iter_snippet = snippet(cx, recv.span, "..");
+        // add note if not multi-line
+        span_lint_and_sugg(
+            cx,
+            CLONED_LAST,
+            expr.span,
+            msg,
+            "try this",
+            format!("{}.last().cloned()", iter_snippet),
```

Comment:
If the `.last()` is `unwrapped` or `?`d, we could even suggest to `.clone()` that. Just a suggestion for extending the lint, this won't stop us from merging.

Reply:
@llogiq Actually, you are right. The result of `last()` is an `Option`, so we should be using `close()` instead.

I updated the PR.

Reply:
@llogiq That's a great suggestion. 

I see:  for `Option<>` /  `Result<>`.
`recv.cloned().unwrap()`
could be
`recv.unwrap().clone`.

I see that `cloned` should be moved furthermost the chain as long as possible, not just for `last`.
For example here:
```rust
 vec.iter().cloned().filter(...).skip(5).take(10).next().last().unwrap()
```
That could be just written as
```rust
let val = vec.cloned().filter(...).skip(5).take(10).next().last().unwrap().clone()
```

This may even allow us to drop the `clone()` entirely, in some cases, when clone() is not needed.

It may be a good idea to propagate clone from before, to after
- `last`
- `take`
- `skip`
- `next`.
- `unwrap` 
- `filter` // has lambda, so requires some more checks
- `find`.// has lambda,so requires some more checks
- ... probably some others as well

The question is, should each be a separate lint, or should they be combined together?

Reply:
I think your example is a bit off, but otherwise full ack.

I'd be OK with multiple lints if there are useful distinctions to make. E.g. if one part of the check can have false positives, it would be a good idea to put that into a different `nursery` lint. Or if one part could reasonably be `style` and another `perf`. Otherwise, putting all the checks into one lint seems natural. One thing I that comes to mind is that we already have the `needless_clone` lint, so we could add cases where the `cloned` can be removed outright to that, making the rest `early_iter_cloned` or some such.

Again, I'm also OK if you restrict the scope of this PR and do a followup later. Have fun!